pgjdbc/src/main/java/org/postgresql/core/QueryExecutorCloseAction.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2023, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.core;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The action performs connection cleanup, so it is properly terminated from the backend
+ * point of view.
+ * Implementation note: it should keep only the minimum number of object references
+ * to reduce heap usage in case the user abandons connection without closing it first.
+ */
+public class QueryExecutorCloseAction implements Closeable {
+  private static final Logger LOGGER = Logger.getLogger(QueryExecutorBase.class.getName());
+  private @Nullable PGStream pgStream;
+
+  public QueryExecutorCloseAction(PGStream pgStream) {
+    this.pgStream = pgStream;
+  }
+
+  public boolean isClosed() {
+    return pgStream == null;
+  }
+
+  public void abort() {
+    PGStream pgStream = this.pgStream;
+    if (pgStream == null) {
+      // The connection has already been closed
+      return;
+    }
+    try {
+      LOGGER.log(Level.FINEST, " FE=> close socket");
+      pgStream.getSocket().close();
+    } catch (IOException e) {
+      // ignore
+    }
+    this.pgStream = null;
+  }
+
+  @Override
+  public void close() throws IOException {
+    LOGGER.log(Level.FINEST, " FE=> Terminate");
+    PGStream pgStream = this.pgStream;
+    if (pgStream == null) {
+      // The connection has already been closed
+      return;
+    }
+    this.pgStream = null;
+    // Technically speaking, this check should not be needed,
+    // however org.postgresql.test.jdbc2.ConnectionTest.testPGStreamSettings
+    // closes pgStream reflectively, so here's an extra check to prevent failures
+    // when getNetworkTimeout is called on a closed stream
+    if (pgStream.isClosed()) {
+      return;
+    }
+    // Prevent blocking the thread for too long
+    // The connection will be discarded anyway, so there's no much sense in waiting long
+    int timeout = pgStream.getNetworkTimeout();
+    if (timeout == 0 || timeout > 1000) {
+      pgStream.setNetworkTimeout(1000);
+    }
+    pgStream.sendChar('X');
+    pgStream.sendInteger4(4);
+
+    pgStream.flush();
+    pgStream.close();
```

Comment:
Ideally, we should move this out of the finalizers thread, as it would get clogged if the application leaks more than one connection per second.
However, I would postpone that to a subsequent PR (e.g. move to PhantomReferences or something like that)

Reply:
leaking more than 1 per second would only be a problem in face of networking issues causing us to hit the timeout, right?
In "happy path" of good connectivity to the database, sending the `X4` should be much faster than that. Or am I missing something?

Reply:
Imagine the user has something like:

```java
for(i in 1..1000) {
  try {var con = openConnection();) {
    con.createStatement().execute(...);
  }
}
// assume the finalizer has not yet executed
// assume the firewall kills the connections, and it blocks any data for them
```

In that case, the finalizer thread would have a pile of 1000 connections to clean up, and it would take ~1000 seconds to unclog while all the other finalizers won't be able to process.

Of course, it is a rare case, however, the finalizer thread is shared across all the other finalizers, so in the ideal world, we should refrain from blocking it even for 1 second.

Reply:
Yeah - this is the horror scenario both in terms of volume of objects created, but also timing of network connectivity loss.